File: src/com/spdqbr/aoc/y2025/Day06.java
package com.spdqbr.aoc.y2025;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

import com.spdqbr.aoc.Solution;

public class Day06 extends Solution {
	public static final int day = 6;
	public static final int year = 2025;
	
	public static void main(String[] args) {
//		DEBUG = true;
		(new Day06()).main(year, day, 0);
		(new Day06()).main(year, day, 1);
		(new Day06()).main(year, day, 2);
		(new Day06()).main(year, day, 3);
	}
	
	public String part1(String[] input) {
		List<Stack<Long>> stacks = new ArrayList<>();
		
		String[] strings;
		long sum = 0;
		for(String s : input) {
			strings = s.trim().split(" +");
			
			for(int i = 0; i < strings.length; i++) {
				if(stacks.size() == i) {
					stacks.add(i, new Stack<Long>());
				}
				if(strings[0].matches("\\d+")) {
					stacks.get(i).add(Long.parseLong(strings[i]));
				} else {
					Long total = null;
					if(strings[i].equals("*")) {
						total = stacks.get(i).pop();
						while(!stacks.get(i).isEmpty()) {
							total *= stacks.get(i).pop();
						}
					}else {
						total = stacks.get(i).pop();
						while(!stacks.get(i).isEmpty()) {
							total += stacks.get(i).pop();
						}
					}
					debug(total);
					sum += total;
				}
			}
		}
		return ""+sum;
	}
	
	public long collapseBuffers(List<StringBuilder> sbs, char op) {
		long total = 0;
		for(int i = 0; i < sbs.size(); i++) {
			if(i == 0) {
				total = Long.parseLong(sbs.get(i).toString());
			}else {
				if(op == '*' ) {
					total *= Long.parseLong(sbs.get(i).toString());
				}else {
					total += Long.parseLong(sbs.get(i).toString());
				}
			}
		}
		debug(total);
		return total;
	}
	
	public String part2(String[] input) {
		long sum = 0;
		List<StringBuilder> sbs = new ArrayList<>();
		int problemCol = 0;
		char op = ' ';
		char c;
		for(int col = 0; col < input[0].length(); col++) {
			boolean allBlank = true;
			for(int row = 0; row < input.length; row ++) {
				c = input[row].charAt(col);
				if(c != ' ') {
					allBlank = false;
					if(sbs.size() == col - problemCol) {
						sbs.add(new StringBuilder());
					}

					if(c != '*' && c != '+'){
						sbs.get(col-problemCol).append(c);
					}else {
						op = c;
					}
				}
			}
			if(allBlank) {
				problemCol=col+1;
				sum += collapseBuffers(sbs, op);
				sbs.clear();
			}
		}
		sum += collapseBuffers(sbs, op);
		
		return ""+sum;
	}
}

File: python/2025/src/Day06.py
from Solution import Solution
import re
import math

part : int = 4

class TodaysSolution(Solution):
    def part_1(self, data: list[str]) -> None:
        s: int = 0

        stacks: list[list[int]] = []
        for row in data:
            vals: list[str] = [x for x in re.split(r' +', row.strip())]
            
            for i in range(len(vals)):
                if len(stacks) <= i:
                    stacks.append([])
                if vals[i] != '*' and vals[i] != '+':
                    stacks[i].append(int(vals[i]))
                else:
                    if vals[i] == '*':
                        val: int = math.prod(stacks[i])
                        print(val)
                        s += val
                    elif vals[i] == '+':
                        print(stacks[i])
                        val: int = sum(stacks[i])
                        print(val)
                        s += val
        print(f"Sum: {s}")
                

    def collapse_buffers(self, buffers: list[list[str]], op: str) -> int:
        vals: list[int] = []
        for buffer in buffers:
            if len(buffer) > 0:
                vals.append(int(''.join(buffer)))
        if op == "*":
            return math.prod(vals)
        return sum(vals)

    def part_2(self, data: list[str]) -> None:
        buffers: list[list[str]] = []

        s: int = 0
        problem_col = 0
        op: str = ' '
        for col in range(len(data[0])):
            all_blank: bool = True
            for row in range(len(data)):
                if len(buffers) <= col-problem_col:
                    buffers.append([])
                if data[row][col] != ' ':
                    all_blank = False
                    if data[row][col] == '*' or data[row][col] == '+':
                        op = data[row][col]
                    else:
                        buffers[col - problem_col].append(data[row][col])
            if all_blank:
                problem_col = col + 1
                s += self.collapse_buffers(buffers, op)
                buffers = []
        
        s += self.collapse_buffers(buffers, op)
        print(f"Sum: {s}")
                
            
def main(filename: str) -> None:
    today: TodaysSolution = TodaysSolution()
    today.solve(filename, part)

if __name__ == '__main__':
    print(__file__)
    main(__file__)